### ISSUE TITLE
Model has 162M parameters, not 125M as expected

### ISSUE THREAD

**Issue by bluecoconut** [KellerJordan/modded-nanogpt#26]
I added this to the code

```py
model = DDP(model, device_ids=[ddp_local_rank])
raw_model = model.module # always contains the "raw" unwrapped model
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print0(f"number of parameters: {trainable_params}")
```

and it prints out:
```
number of parameters: 162201636
```

checking the math, it seems right
```
Embedding Layer: 
  Total: 50,304 × 768 = 38,633,472
Transformer Blocks
  Block:
    - Attention:
        - LinearLayers: 4 × (768 × 768) = 2,359,296
        - Lambda: 1
        Total: 2,359,296 + 1 = 2,359,297
    - MLP:
        Total: 2 × (768 × 3,072) = 4,718,592
    - Lambdas: 2
    Total: 2,359,297 + 4,718,592 + 2 = 7,077,891
  Total: 12 × 7,077,891 = 84,934,692
LM-Head:
  Total: 768 × 50,304 = 38,633,472
Total: 38,633,472 + 84,934,692 + 38,633,472 = 162,201,636
```

The addition of a trainable `lm_head` rather than sharing the same weights as the embedding layer added an extra ~40M parameters led to this increase. In order to keep 125M parameters (to match) need to go down to ~7 layers (126.8M) or 6 layers (119.7M).

In an effort to see how much impact this has, I did a run (with the unet skips) and 6 layers for 5000 steps -> step_avg time is down to 93.53 ms (due to 6 layers). However, at equal time run (~8 min) it's not getting to the same loss / keeping up with the larger model.

![image](https://github.com/user-attachments/assets/42bad459-8d22-4783-9a59-bf83422b242c)
![image](https://github.com/user-attachments/assets/54b07306-0d16-40ff-9105-2ddb9078b0b6)

**Comment by bluecoconut** [comment]
From X:
https://x.com/kellerjordan0/status/1856501669510688968
> True, they have 162M params, but 38M of that is embedding, of which only 768 params are active per token
